plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,102 +64,126 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
-                            }
-                            if (range.isSingleValue()) {
-                                singleValues.add(columnHandle.getCassandraType().toCqlLiteral(range.getSingleValue()));
-                            }
-                            else {
-                                if (!range.isLowUnbounded()) {
-                                    String lowBound = columnHandle.getCassandraType().toCqlLiteral(range.getLowBoundedValue());
-                                    rangeConjuncts.add(format(
-                                            "%s %s %s",
-                                            CassandraCqlUtils.validColumnName(columnHandle.getName()),
-                                            range.isLowInclusive() ? ">=" : ">",
-                                            lowBound));
-                                }
-                                if (!range.isHighUnbounded()) {
-                                    String highBound = columnHandle.getCassandraType().toCqlLiteral(range.getHighBoundedValue());
-                                    rangeConjuncts.add(format(
-                                            "%s %s %s",
-                                            CassandraCqlUtils.validColumnName(columnHandle.getName()),
-                                            range.isHighInclusive() ? "<=" : "<",
-                                            highBound));
-                                }
-                            }
-                        }
-
-                        if (!singleValues.isEmpty() && !rangeConjuncts.isEmpty()) {
+                        if (ranges.getSpan().isAll()) {
                             return null;
                         }
-                        if (!singleValues.isEmpty()) {
-                            if (singleValues.size() == 1) {
-                                predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = " + singleValues.get(0);
-                            }
-                            else {
-                                predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
-                                        + Joiner.on(",").join(singleValues) + ")";
-                            }
+                        if (ranges.getRangeCount() == 1) {
+                            fullyPushedDomains.add(columnHandle);
+                            return translateRangeIntoCql(columnHandle, ranges.getOrderedRanges().get(0));
                         }
-                        else if (!rangeConjuncts.isEmpty()) {
-                            predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                        if (ranges.getOrderedRanges().stream().allMatch(Range::isSingleValue)) {
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                return null;
```

Comment:
You could still do `return translateRangeIntoCql(columnHandle, ranges.getSpan());` (as below)